### PR TITLE
Fix config repository rules violation (#8368)

### PR DESCRIPTION
**Pull request by GaneshSPatil**
Issue: #8368

Description:

* Remove pipelines from config on rules failure.
* Remove cached partial when merge fails.
* Revalidate last valid partials with latest config repo rules.

### CODE REVIEW

**Comment by GaneshSPatil** [comment]
PR Build: https://build.gocd.org/go/materials/value_stream_map/b8b134da75623e1311ef993a270b8c4fcced32ce53fa84072f6cb3f9a348a1a6/dc5a9bb3591333ee12d28991356a4791b31e4185

File: server/src/main/java/com/thoughtworks/go/config/PartialConfigService.java [hunk 1]
```diff
@@ -69,27 +70,57 @@ public void onSuccessPartialConfig(ConfigRepoConfig repoConfig, PartialConfig in
...
+                        if (isSameRevisionAsOfPreviousValidPartial) {
+                            removeCachedPartialsAndUpdateConfig(fingerprint);
+                        } else {
+                            // lets say the latest partial is invalid for the current config repo rules.
+                            // in such cases, validate the previous valid partial with respect to current config repo rules.
+                            // if the previous valid partials are valid - do nothing - as the error for the latest partial is already populated and config contains the last known partial
+                            // and if the previous valid partials are invalid - remove those config without clearing the server health message. Server health message is populated for the same fingerprint with the latest parse failure message.
+                            ((RepoConfigOrigin) previousValidPartial.getOrigin()).setConfigRepo(((RepoConfigOrigin) incoming.getOrigin()).getConfigRepo());
+                            if (hasRuleViolations(previousValidPartial)) {
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/config/PartialConfigService.java:102]: I get your point why we need to check rules for last valid partial against the current rules. guess my previous comment has to change to something like,

```
if (updateConfig(incoming, fingerprint, repoConfig)) {
    cachedGoPartials.markAsValid(fingerprint, incoming);
} else {
    final PartialConfig previousValidPartial = cachedGoPartials.getValid(repoConfig.getRepo().getFingerprint());
    if (validPartialHasRuleViolations(previousValidPartial)) {
        // do not allow fallback to the last version of the partial if the current rules do not allow
        cachedGoPartials.removeValidWithoutClearingServerHealthMessage(repoConfig.getRepo().getFingerprint());
        goConfigService.updateConfig(cruiseConfig -> {
            cruiseConfig.getPartials().remove(cachedGoPartials.findPartialByFingerprint(cruiseConfig, fingerprint));
            return cruiseConfig;
        });
    }
}
```

**Comment by GaneshSPatil** [comment]
DNM -- requires changes

File: server/src/main/java/com/thoughtworks/go/config/PartialConfigService.java [hunk 2]
```diff
@@ -69,21 +70,40 @@ public void onSuccessPartialConfig(ConfigRepoConfig repoConfig, PartialConfig in
...
-                // rule violations will be considered before accepting a merge;
-                // updateConfig() only considers structural validity.
-                final boolean violatesRules = hasRuleViolations(incoming);
-
+                // mark the fingerprint as last known
                 cachedGoPartials.cacheAsLastKnown(fingerprint, incoming);
 
-                if (updateConfig(incoming, fingerprint, repoConfig)) {
+                //validate rules
+                hasRuleViolations(incoming);
```
Comment (maheshp) [inline server/src/main/java/com/thoughtworks/go/config/PartialConfigService.java:77]: @GaneshSPatil do we need to explicitly check for rules violation here? Wouldn't config update apply rules?

**Comment by maheshp** [comment]
Verified on GoCD Version: 20.6.0 (11995-0500dc2e9647d2ccffb976c3d9228bad7426fdbd).

**Pull request merged**